Keystone connector

kbalka opened the pull request:
PR contains connector for openstack keystone.

Features:
- access tokens
- refresh tokens 
- groups

Requirements:
- access to openstack keystone instance
- keystone administrative account credentials 

Enabling keystone connector specific tests:
- make sure docker is running 
- export DEX_TEST_KEYSTONE=1
- make tests

srenatus commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,158 @@
+	// Providing wrong password or wrong keystone URI throws error
+	if err == nil && response.StatusCode == 201 {
+    	token := response.Header["X-Subject-Token"][0]
+		data, _ := ioutil.ReadAll(response.Body)
```
Let's please check all errors (same applies to other calls in this PR) -- it's fine to skip an error check or two in the test files, but let's please not do that here.

srenatus commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,158 @@
+	response, err := p.getTokenResponse(username, password)
+
+	// Providing wrong password or wrong keystone URI throws error
+	if err == nil && response.StatusCode == 201 {
```
Let's stick to idiomatic Go code, like
```go
resp, err := p.getTokenResponse(username, password)
if err != nil {
    // handle err
}
if resp.StatusCode == 201 {
    // ... read body
```

Also, `resp.Body` will need to be closed, otherwise we risk leaking memory.

srenatus commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,158 @@
+
+func (p keystoneConnector) Login(ctx context.Context, s connector.Scopes, username, password string) (
+		identity connector.Identity, validPassword bool, err error) {
+	response, err := p.getTokenResponse(username, password)
```
It's better to thread the `ctx` parameter through here, and use it when making the request to Keystone -- see [the github connector for example](https://github.com/dexidp/dex/blob/4329f407dcfefad56aa81b4d3da1a8e487bd023e/connector/github/github.go#L483-L488).

srenatus commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,158 @@
+
+	// Providing wrong password or wrong keystone URI throws error
+	if err == nil && response.StatusCode == 201 {
+    	token := response.Header["X-Subject-Token"][0]
```
[nit] We could use [`(*Header).Get()`](https://golang.org/pkg/net/http/#Header.Get) here to not depend on the case of the header key.

srenatus commented on connector/keystone/types.go:
```diff
@@ -0,0 +1,154 @@
+	Name string `json:"name"`
+}
+
+// Config holds the configuration parameters for Keystone connector.
```
Do you know which (API) version of keystone we're requiring here? There's a few https://developer.openstack.org /api-ref/identity/ 😉

kbalka replied:
We are requiring API version 3. Could you have a second look at this PR ? It seems that CI builds are passing now ;-)

srenatus commented on connector/keystone/types.go:
```diff
@@ -0,0 +1,154 @@
+	Token token `json:"token"`
+}
+
+type createUserRequest struct {
```
Can we put the definitions which we only need for tests into the test file, please?

srenatus commented on examples/config-keystone.yaml:
```diff
@@ -0,0 +1,55 @@
+# The base path of dex and the external name of the OpenID Connect service.
```
I don't think we need one of these example files for every connector. Let's remove it. 😃

srenatus commented on connector/keystone/keystone_test.go:
```diff
@@ -0,0 +1,326 @@
+
+const domain = "default"
+
+func startKeystoneContainer() string {
```
This is a nice integration test suite.

But let's thread in a `t *testing.T` argument use `t.Helper()` and `t.Fatal` in the helper methods, instead of `fmt.Printf` calls, please 😉

srenatus commented on connector/keystone/keystone_test.go:
```diff
@@ -0,0 +1,326 @@
+  	}
+  	defer cleanKeystoneContainer(dockerID)
+  	// run all tests
+	m.Run()
```
❓ Do we need `os.Exit(m.Run())` here to have the test run fail when one of the tests fails?

srenatus commented on connector/keystone/keystone_test.go:
```diff
@@ -0,0 +1,360 @@
+)
+
+const (
+	adminUser   = "demo"
```
❓ Would it make sense to pass these in, too, at least in so far as they're specific to https://hub.docker.com/r/openio/openstack-keystone?

srenatus commented on .travis.yml:
```diff
@@ -13,13 +13,14 @@ services:
   - sudo /etc/init.d/slapd stop
   - docker run -d --net=host gcr.io/etcd-development/etcd:v3.2.9
-
+  - docker run -d -p 0.0.0.0:5000:5000 -p 0.0.0.0:35357:35357 openio/openstack-keystone
```
❓ Is there any version pins we could add? Right now, this will take the "latest" tag, and that'll potentially break a month (or a year) from now, and will cause some PR to go ":x:" for the wrong reasons :wink:

kbalka replied:
This image has only "latest" tag, so the only thing we can do here is to fork it on dockerhub and add specific tag.

srenatus replied:
👍 Let's be optimistic. We can still fork it if the need arises down the road. (But really, I've rather avoid that 😅)

ericchiang replied:
Is there an official image for Keystone? I don't trust a "latest" tag from a organization that doesn't maintain the project.

kbalka replied:
I could not find an "official" image for Keystone. I've asked the maintainer of the image to add tag, so right now we are using "pike" tag. Is this good enough ?

srenatus replied:
I think it's a good compromise. Thanks for engaging with the maintainer there 👍

srenatus commented on .travis.yml:
```diff
@@ -13,13 +13,14 @@ services:
   - docker run -d --net=host gcr.io/etcd-development/etcd:v3.2.9
-
+  - docker run -d -p 0.0.0.0:5000:5000 -p 0.0.0.0:35357:35357 openio/openstack-keystone
+  - sleep 60s
```
🤔 Can we do something that is a bit more robust? 
Maybe (untested)
```yaml
  - |
    until curl --fail http://127.0.0.1:5000/v3; do
      echo 'Waiting for keystone...'
      sleep 1;
    done;
```
This would potentially be faster (wait for less than 60s) -- OTOH, it could take longer (if keystone does not come up), but I'd hope travis will fail the build with some timeout in that case.

Either way, this new additional dependency in our test suite is something to keep an eye on when this is merged. 🔍

srenatus commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,189 @@
+			return identity, false, err
+		}
+
+		var tokenResponse = new(tokenResponse)
```
[nit] I'm not a fan of clashing names between types and variables; but if you are, that's tine 😉

srenatus commented:
🎗 Note to self: *let's squash these before merging*.

srenatus commented:
Not right now, though! I appreciate the separate commits a lot for review 👍

joannanosek commented:
Hi @srenatus! Sorry for torpedoing with the commits 😅 And thank you for valuable feedback!

kbalka commented on .travis.yml:
```diff
@@ -13,13 +13,14 @@ services:
   - sudo /etc/init.d/slapd stop
   - docker run -d --net=host gcr.io/etcd-development/etcd:v3.2.9
-
+  - docker run -d -p 0.0.0.0:5000:5000 -p 0.0.0.0:35357:35357 openio/openstack-keystone
```
Good question :-) For me, the advantage here is only readability - you can see which ports are exposed and compare it with exported variables DEX_KEYSTONE*.

kbalka commented:
the application which can possibly leverage openstack cluster authentication is https://github.com/IntelAI/inference-model-manager :-)

kbalka commented on connector/keystone/types.go:
```diff
@@ -0,0 +1,102 @@
+	Methods   []string               `json:"methods"`
+	ExpiresAt string                 `json:"expires_at"`
+	User      userKeystone           `json:"user"`
+}
```
I think we don't need to care about ExpiresAt - because we are using token immediately after it's obtained. Keystone tokens are quite short living (I think by default 1h). Keeping it in identity.ConnectorData and using for Refresh would limit liveness of dex refresh token by expiration of keystone user token. That's why we require keystone admin account credentials. Admin should have access to other's users and their groups. When we check refresh token, the admin account gets fresh token from keystone and verifies user existence and groups. I hope that it makes some sense :-)

ericchiang commented on connector/keystone/types.go:
```diff
@@ -0,0 +1,87 @@
+package keystone
```
Please don't separate types and methods. These should all go in the same file

ericchiang commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,187 @@
+}
+
+func (p *keystoneConnector) checkIfUserExists(ctx context.Context, userID string, token string) (bool, error) {
+	userURL := p.KeystoneHost + "/v3/users/" + userID
```
please link to upstream docs describing this endpoint

ericchiang commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,187 @@
+func (p *keystoneConnector) Refresh(
+	ctx context.Context, s connector.Scopes, identity connector.Identity) (connector.Identity, error) {
+
+	token, err := p.getAdminToken(ctx)
```
Should this token be cached?

kbalka replied:
We can add caching token across sessions if needed, but my opinion is that at this moment we should keep the code simple. Is it ok for you ?

ericchiang commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,187 @@
+		return identity, fmt.Errorf("keystone: user %q does not exist", identity.UserID)
+	}
+
+	groups, err := p.getUserGroups(ctx, identity.UserID, token)
```
groups should only be set if the scope is requested

ericchiang commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,187 @@
+	}
+
+	// Providing wrong password or wrong keystone URI throws error
+	if resp.StatusCode == 201 {
```
Needs to check for non 2xx status codes. Also, just return early:

```
if resp.StatusCode / 100 != 2 {
    // return error describing status code
}
if resp.StatusCode != 201 {
    return identity, false, nil
}
// other logic
```

ericchiang commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,187 @@
+func (p *keystoneConnector) Close() error { return nil }
+
+func (p *keystoneConnector) Login(ctx context.Context, s connector.Scopes, username, password string) (
+	identity connector.Identity, validPassword bool, err error) {
```
nit: keep these arguments on one line

ericchiang commented on connector/keystone/types.go:
```diff
@@ -0,0 +1,87 @@
+	"github.com/sirupsen/logrus"
+)
+
+type keystoneConnector struct {
```
already in the keystone package, so this can just be called "connector"

srenatus replied:
How about `conn` or plain `c`? Currently, other connectors use a similar name to what's introduced here (e.g. `microsoftConnector`), but I do agree that shorter is nicer, the context is clear here.

kbalka replied:
Sure, it makes sense. I've changed to `conn`.

ericchiang commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,187 @@
+		if err != nil {
+			return identity, false, fmt.Errorf("keystone: invalid token response: %v", err)
+		}
+		groups, err := p.getUserGroups(ctx, tokenResp.Token.User.ID, token)
```
Groups should only be fetched if the scope is requested

kbalka replied:
Good point, I've added a logic and test case

ericchiang commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,187 @@
+
+// Open returns an authentication strategy using Keystone.
+func (c *Config) Open(id string, logger logrus.FieldLogger) (connector.Connector, error) {
+	return &keystoneConnector{c.Domain, c.KeystoneHost,
```
nit: format this with fields on different lines

```
return &keystoneConnector{
    c.Doman,
    c.KeystoneHost,
    c.KeystoneUsername,
    c.KeystonePassword,
    logger,
}
```

ericchiang commented on connector/keystone/types.go:
```diff
@@ -0,0 +1,87 @@
+
+type keystoneConnector struct {
+	Domain           string
+	KeystoneHost     string
```
These should be `Host`, `AdminUsername`, `AdminPassword`, right?

ericchiang commented on connector/keystone/types.go:
```diff
@@ -0,0 +1,87 @@
+//      keystonePassword: DEMO_PASS
+type Config struct {
+	Domain           string `json:"domain"`
+	KeystoneHost     string `json:"keystoneHost"`
```
Yeah, "host", "username", "password"

kbalka commented:
Hi @ericchiang, thanks for a thorough review. I'd addressed most of the issues. Please check if it looks better now. Regarding use of openio/openstack-keystone:latest image, I've asked the maintainer to add specific version tag, waiting for response. But, the most important, merry 🎄 and 🎅 to you together with @joannanosek @srenatus and @soggiest !

soggiest commented:
Happy Holidays and New Years!

On Fri, Dec 21, 2018 at 7:15 AM Krzysztof Balka <notifications@github.com>
wrote:

> Hi @ericchiang <https://github.com/ericchiang>, thanks for a thorough
> review. I'd addressed most of the issues. Please check if it looks better
> now. Regarding use of openio/openstack-keystone:latest image, I've asked
> the maintainer to add specific version tag, waiting for response. But, the
> most important, merry 🎄 and 🎅 to you together with @joannanosek
> <https://github.com/joannanosek> @srenatus <https://github.com/srenatus>
> and @soggiest <https://github.com/soggiest> !
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dexidp/dex/pull/1374#issuecomment-449414066>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGDIln8q6NvUwC2f2I-uTo73_AtOsdfnks5u7PsYgaJpZM4ZTvYH>
> .
>

kbalka commented on connector/keystone/keystone.go:
```diff
@@ -0,0 +1,189 @@
+	return identity, false, nil
+}
+
+func (p keystoneConnector) Prompt() string { return "username" }
```
I know it can be integrated with LDAP and ActiveDirectory.

kbalka commented:
Hi @srenatus. Thanks for the approval !!! I've squashed the commits into number of 4, added authors and force-pushed the branch on my fork. The code seems to be ready for merge. The only strange thing I can observe that there is a "tail" of 8 commits in the PR which are already on the master. They will probably disappear after merge, but I'm not sure. I can fix the PR if needed.

srenatus commented:
They should disappear if you fetch origin and rebase upon origin/master :)

I'd prefer getting this straightened out before hitting the button...

On Fri, 11 Jan 2019, 13:45 Krzysztof Balka <notifications@github.com wrote:

> Hi @srenatus <https://github.com/srenatus>. Thanks for the approval !!!
> I've squashed the commits into number of 4, added authors and force-pushed
> the branch on my fork. The code seems to be ready for merge. The only
> strange thing I can observe that there is a "tail" of 8 commits in the PR
> which are already on the master. They will probably disappear after merge,
> but I'm not sure. I can fix the PR if needed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dexidp/dex/pull/1374#issuecomment-453506069>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1I7vdXcv2F5Tbd0DCcVzoRgFw0lZEAks5vCId7gaJpZM4ZTvYH>
> .
>

srenatus commented:
@kbalka Thank you so much! 🎉 

@ericchiang sorry I didn't wait for your "👍" after all. Feel free to assign me anything that's left 😉
